File: bin/plot_comptime.py
#!/usr/bin/env python3
"""
Computational Performance Figure (Supplementary S1)

Creates horizontal bar charts showing:
- Mean runtime by process
- Memory usage as secondary axis or separate panel

Cleaner than previous violin plot approach.

Usage:
    python plot_comptime.py \
        --all_runs path/to/runs_directory \
        --outdir figures
"""

import argparse
import os
import re
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import yaml

# Import shared utilities
from plot_utils import (
    set_pub_style, save_figure,
    METHOD_COLORS,
    SINGLE_COL, HALF_HEIGHT, FULL_WIDTH, STANDARD_HEIGHT
)


# Process display names and colors
# Keys are the final component of the Nextflow process name (after last ':')
PROCESS_NAMES = {
    'RF_PREDICT': 'scVI Predict',
    'PREDICT_SEURAT': 'Seurat Predict',
    'MAP_QUERY': 'scVI Query',
    'QUERY_PROCESS_SEURAT': 'Seurat Query',
    'REF_PROCESS_SEURAT': 'Seurat Ref'
}

PROCESS_COLORS = {
    'RF_PREDICT': '#1f77b4',
    'PREDICT_SEURAT': '#ff7f0e',
    'MAP_QUERY': '#2ca02c',
    'QUERY_PROCESS_SEURAT': '#d62728',
    'REF_PROCESS_SEURAT': '#9467bd'
}

# Process ordering (prediction methods first, then preprocessing)
PROCESS_ORDER = [
    'RF_PREDICT', 'PREDICT_SEURAT',
    'MAP_QUERY', 'QUERY_PROCESS_SEURAT', 'REF_PROCESS_SEURAT'
]


def parse_arguments():
    parser = argparse.ArgumentParser(
        description="Generate computational performance figures"
    )
    parser.add_argument(
        '--all_runs', type=str, required=True,
        help="Path to directory containing trace results"
    )
    parser.add_argument(
        '--outdir', type=str, default='.',
        help="Output directory for figures"
    )
    parser.add_argument(
        '--output_prefix', type=str, default='comptime',
        help="Prefix for output filenames"
    )

    if __name__ == "__main__":
        return parser.parse_args()
    return parser.parse_known_args()[0]


def convert_time(time_str) -> float:
    """Convert time string to hours."""
    if pd.isna(time_str) or time_str == "-" or not isinstance(time_str, str):
        return np.nan

    time_str = time_str.lower().replace(" ", "")

    # Match hours, minutes, seconds, milliseconds
    match = re.match(r"(?:(\d+)h)?(?:(\d+)m)?(?:(\d+\.?\d*)s)?(?:(\d+)ms)?", time_str)

    if not match:
        return np.nan

    hours = int(match.group(1)) if match.group(1) else 0
    minutes = int(match.group(2)) if match.group(2) else 0
    seconds = float(match.group(3)) if match.group(3) else 0
    milliseconds = float(match.group(4)) / 1000 if match.group(4) else 0

    total_seconds = hours * 3600 + minutes * 60 + seconds + milliseconds
    return total_seconds / 3600  # Return hours


def convert_percent(value) -> float:
    """Convert percentage string to float."""
    if isinstance(value, str) and value.endswith("%"):
        return float(value.strip("%"))
    return np.nan


def convert_memory(value_str) -> float:
    """Convert memory string to GB."""
    if pd.isna(value_str) or value_str == "-":
        return np.nan

    if isinstance(value_str, str):
        parts = value_str.strip().split()
        if len(parts) >= 1:
            try:
                val = float(parts[0])
                # Assume MB if no unit, convert to GB
                return val / 1024
            except ValueError:
                return np.nan
    return np.nan


def load_trace_data(all_runs_dir: str) -> pd.DataFrame:
    """Load and combine all trace files found recursively."""
    reports = pd.DataFrame()

    for root, dirs, files in os.walk(all_runs_dir):
        if "trace.txt" not in files:
            continue

        trace_path = os.path.join(root, "trace.txt")
        params_path = os.path.join(root, "params.yaml")

        try:
            trace = pd.read_csv(trace_path, sep="\t")

            # Load params if available
            if os.path.exists(params_path):
                with open(params_path, "r") as f:
                    params = yaml.safe_load(f)

                # Remove non-scalar params
                keys_to_drop = ["ref_collections", "ref_keys", "outdir",
                               "batch_keys", "relabel_r", "relabel_q",
                               "tree_file", "queries_adata"]
                for key in keys_to_drop:
                    params.pop(key, None)

                for key, value in params.items():
                    trace[key] = value

            reports = pd.concat([reports, trace], ignore_index=True)

        except Exception as e:
            print(f"Warning: Could not load {trace_path}: {e}")

    return reports


def create_line_plot(
    ax: plt.Axes,
    stats: pd.DataFrame,
    mean_col: str,
    std_col: str,
    ylabel: str,
    process_order: list = None
) -> plt.Axes:
    """
    Create a line/point plot with error bars for process steps on the x-axis.
    """
    if process_order is None:
        process_order = stats.index.tolist()

    available = [p for p in process_order if p in stats.index]
    plot_data = stats.loc[available]

    positions = np.arange(len(available))
    colors = [PROCESS_COLORS.get(p, '#666666') for p in available]
    labels = [PROCESS_NAMES.get(p, p) for p in available]

    # Plot connecting line
    ax.plot(positions, plot_data[mean_col], color='grey', linewidth=1, zorder=1)

    # Plot points with error bars per process
    for i, (proc, pos) in enumerate(zip(available, positions)):
        ax.errorbar(
            pos, plot_data.loc[proc, mean_col],
            yerr=plot_data.loc[proc, std_col],
            fmt='o', color=colors[i], markersize=8,
            capsize=4, capthick=1.5, elinewidth=1.5,
            markeredgecolor='white', markeredgewidth=0.5,
            zorder=2
        )

    ax.set_xticks(positions)
    ax.set_xticklabels(labels, rotation=45, ha='right')
    ax.set_ylabel(ylabel)
    ax.set_ylim(0, None)

    return ax


def main():
    args = parse_arguments()

    # Set publication style
    set_pub_style()

    # Create output directory
    os.makedirs(args.outdir, exist_ok=True)

    # Load data
    print("Loading trace data...")
    reports = load_trace_data(args.all_runs)

    if len(reports) == 0:
        print("Error: No trace data found")
        return

    # Process data
    print("Processing data...")

    # Convert columns
    if '%cpu' in reports.columns:
        reports['%cpu'] = reports['%cpu'].apply(convert_percent)
    if 'duration' in reports.columns:
        reports['duration_hours'] = reports['duration'].apply(convert_time)
    if 'realtime' in reports.columns:
        reports['realtime_hours'] = reports['realtime'].apply(convert_time)
    if 'peak_vmem' in reports.columns:
        reports['memory_gb'] = reports['peak_vmem'].apply(convert_memory)

    # Extract process name: "SEURAT_PIPELINE:PREDICT_SEURAT (5)" -> "PREDICT_SEURAT"
    reports['process'] = reports['name'].apply(
        lambda x: str(x).split('(')[0].strip().split(':')[-1] if pd.notna(x) else None
    )

    # Filter to relevant processes
    processes = [p for p in PROCESS_ORDER if p in reports['process'].values]
    trace_subset = reports[reports['process'].isin(processes)].copy()

    if len(trace_subset) == 0:
        print("Error: No relevant processes found in trace data")
        return

    # Calculate statistics
    stats = trace_subset.groupby('process').agg({
        'duration_hours': ['mean', 'std', 'count'],
        'realtime_hours': ['mean', 'std'],
        'memory_gb': ['mean', 'std']
    })

    # Flatten column names
    stats.columns = ['mean_duration', 'std_duration', 'count',
                     'mean_realtime', 'std_realtime',
                     'mean_memory', 'std_memory']

    # Fill NaN std with 0
    stats = stats.fillna(0)

    print(f"Found {len(stats)} processes with data")
    print(stats)

    # Save summary table
    stats_out = stats.copy()
    stats_out.index = stats_out.index.map(lambda x: PROCESS_NAMES.get(x, x))
    stats_out.to_csv(os.path.join(args.outdir, f"{args.output_prefix}_summary.tsv"), sep='\t')

    # Create figure with two stacked panels
    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(FULL_WIDTH, STANDARD_HEIGHT))

    # Panel A: Runtime
    create_line_plot(ax1, stats, 'mean_duration', 'std_duration',
                     'Duration (hours)', PROCESS_ORDER)
    ax1.set_title('A. Runtime', fontweight='bold', loc='left')

    # Panel B: Memory
    create_line_plot(ax2, stats, 'mean_memory', 'std_memory',
                     'Peak Memory (GB)', PROCESS_ORDER)
    ax2.set_title('B. Peak Memory', fontweight='bold', loc='left')

    fig.tight_layout(h_pad=3)

    # Save
    output_path = os.path.join(args.outdir, args.output_prefix)
    print(f"Saving figure to {output_path}...")
    save_figure(fig, output_path, formats=['pdf', 'png'], dpi=300)

    plt.close(fig)
    print("Done!")


if __name__ == "__main__":
    main()

File: bin/plot_comptime.R
#!/usr/bin/env Rscript
# Computational Performance Figure (Supplementary S1)
#
# Creates point-and-line plots with error bars showing:
#   Panel A: Mean runtime (hours) by process step
#   Panel B: Mean peak memory (GB) by process step
#
# Usage:
#   Rscript plot_comptime.R --all_runs path/to/runs_directory --outdir figures

library(argparse)
library(dplyr)
library(tidyr)
library(readr)
library(stringr)
library(ggplot2)
library(patchwork)
library(yaml)

# -- Constants ----------------------------------------------------------------

# Map each Nextflow process to its methodology and pipeline stage
PROCESS_METHOD <- c(
  MAP_QUERY            = "scVI",
  RF_PREDICT           = "scVI",
  REF_PROCESS_SEURAT   = "Seurat",
  QUERY_PROCESS_SEURAT = "Seurat",
  PREDICT_SEURAT       = "Seurat"
)

PROCESS_STEP <- c(
  MAP_QUERY            = "Query Processing",
  RF_PREDICT           = "Prediction",
  REF_PROCESS_SEURAT   = "Ref Processing",
  QUERY_PROCESS_SEURAT = "Query Processing",
  PREDICT_SEURAT       = "Prediction"
)

# Pipeline stage order (processing → prediction)
STEP_ORDER <- c("Ref Processing", "Query Processing", "Prediction")

# Method colors (consistent with plot_utils.py METHOD_COLORS)
METHOD_COLORS <- c(scVI = "#1f77b4", Seurat = "#ff7f0e")

PROCESS_ORDER <- names(PROCESS_METHOD)

KEYS_TO_DROP <- c("ref_collections", "ref_keys", "outdir",
                  "batch_keys", "relabel_r", "relabel_q",
                  "tree_file", "queries_adata")

# -- Helpers ------------------------------------------------------------------

convert_time <- function(x) {
  # Convert Nextflow time strings (e.g. "1h30m45.2s", "200ms") to hours
  ifelse(is.na(x) | x == "-", NA_real_, {
    x <- tolower(gsub(" ", "", x))
    hrs  <- as.numeric(ifelse(grepl("(\\d+)h", x), sub(".*?(\\d+)h.*", "\\1", x), "0"))
    mins <- as.numeric(ifelse(grepl("(\\d+)m(?!s)", x, perl = TRUE),
                              sub(".*?(\\d+)m(?!s).*", "\\1", x, perl = TRUE), "0"))
    secs <- as.numeric(ifelse(grepl("([\\d.]+)s(?!.*ms)", x, perl = TRUE),
                              sub(".*?([\\d.]+)s.*", "\\1", x, perl = TRUE), "0"))
    ms   <- as.numeric(ifelse(grepl("(\\d+)ms", x), sub(".*?(\\d+)ms.*", "\\1", x), "0"))
    (hrs * 3600 + mins * 60 + secs + ms / 1000) / 3600
  })
}

convert_memory <- function(x) {
  # Convert memory string (first token, assumed MB) to GB
  ifelse(is.na(x) | x == "-", NA_real_, {
    val <- suppressWarnings(as.numeric(str_extract(x, "[\\d.]+")))
    val / 1024
  })
}

# -- Data loading -------------------------------------------------------------

load_trace_data <- function(all_runs_dir) {
  trace_files <- list.files(all_runs_dir, pattern = "^trace\\.txt$",
                            recursive = TRUE, full.names = TRUE)
  if (length(trace_files) == 0) return(tibble())

  traces <- lapply(trace_files, function(tf) {
    tryCatch({
      trace <- read_tsv(tf, show_col_types = FALSE)
      params_path <- file.path(dirname(tf), "params.yaml")
      if (file.exists(params_path)) {
        params <- yaml::read_yaml(params_path)
        params[KEYS_TO_DROP] <- NULL
        # Keep only scalar params
        params <- params[vapply(params, function(v) length(v) == 1 && !is.list(v), logical(1))]
        for (nm in names(params)) trace[[nm]] <- params[[nm]]
      }
      trace
    }, error = function(e) {
      message("Warning: Could not load ", tf, ": ", conditionMessage(e))
      NULL
    })
  })

  bind_rows(traces)
}

# -- Plotting -----------------------------------------------------------------

make_panel <- function(stats_df, mean_col, sd_col, ylabel, title_label,
                       show_legend = FALSE) {
  has_subsample <- "subsample_ref" %in% names(stats_df)
  group_col <- if (has_subsample) {
    interaction(stats_df$method, as.character(stats_df$subsample_ref), drop = TRUE)
  } else {
    stats_df$method
  }

  p <- ggplot(stats_df, aes(x = step, y = .data[[mean_col]],
                             color = method, group = group_col)) +
    geom_line(aes(linetype = if (has_subsample) factor(subsample_ref) else NULL),
              linewidth = 0.7) +
    geom_pointrange(
      aes(ymin = .data[[mean_col]] - .data[[sd_col]],
          ymax = .data[[mean_col]] + .data[[sd_col]]),
      size = 0.6, linewidth = 0.8
    ) +
    scale_color_manual(values = METHOD_COLORS, name = "Method") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05)), limits = c(0, NA)) +
    labs(y = ylabel, x = NULL, title = title_label) +
    theme_bw(base_size = 11) +
    theme(
      plot.title = element_text(face = "bold", hjust = 0),
      axis.text.x = element_text(angle = 45, hjust = 1),
      plot.background  = element_rect(fill = "white", color = NA),
      panel.background = element_rect(fill = "white", color = NA),
      panel.grid.major.x = element_blank(),
      legend.position = if (show_legend) "right" else "none"
    )

  if (has_subsample) {
    p <- p + scale_linetype_discrete(name = "Subsample Ref")
  }

  p
}

# -- Main ---------------------------------------------------------------------

main <- function() {
  parser <- ArgumentParser(description = "Generate computational performance figures")
  parser$add_argument("--all_runs", required = TRUE, help = "Path to runs directory")
  parser$add_argument("--outdir", default = ".", help = "Output directory")
  parser$add_argument("--output_prefix", default = "comptime", help = "Output filename prefix")
  args <- parser$parse_args()

  dir.create(args$outdir, showWarnings = FALSE, recursive = TRUE)

  cat("Loading trace data...\n")
  reports <- load_trace_data(args$all_runs)
  if (nrow(reports) == 0) { stop("No trace data found") }

  # Convert columns
  if ("duration"  %in% names(reports)) reports$duration_hours <- convert_time(reports$duration)
  if ("realtime"  %in% names(reports)) reports$realtime_hours <- convert_time(reports$realtime)
  if ("peak_vmem" %in% names(reports)) reports$memory_gb     <- convert_memory(reports$peak_vmem)

  # Extract process name: "PIPELINE:PROCESS_NAME (sample)" -> "PROCESS_NAME"
  reports$process <- str_trim(str_extract(reports$name, "[^:]+$")) |>
    str_remove("\\s*\\(.*\\)$")

  # Filter to relevant processes
  reports <- reports %>% filter(process %in% PROCESS_ORDER)
  if (nrow(reports) == 0) stop("No relevant processes found in trace data")

  # Add method and pipeline step
  reports$method <- PROCESS_METHOD[reports$process]
  reports$step   <- PROCESS_STEP[reports$process]

  # Summarise per process (optionally by subsample_ref)
  group_cols <- c("process", "method", "step")
  if ("subsample_ref" %in% names(reports)) {
    group_cols <- c(group_cols, "subsample_ref")
  }

  stats <- reports %>%
    group_by(across(all_of(group_cols))) %>%
    summarise(
      mean_duration = mean(duration_hours, na.rm = TRUE),
      sd_duration   = sd(duration_hours,   na.rm = TRUE),
      mean_realtime = mean(realtime_hours,  na.rm = TRUE),
      sd_realtime   = sd(realtime_hours,    na.rm = TRUE),
      mean_memory   = mean(memory_gb,       na.rm = TRUE),
      sd_memory     = sd(memory_gb,         na.rm = TRUE),
      count         = n(),
      .groups = "drop"
    ) %>%
    mutate(across(where(is.numeric), ~ replace_na(., 0)))

  # Ordered factor for x-axis (pipeline order: processing → prediction)
  stats$step <- factor(stats$step, levels = STEP_ORDER)

  cat("Found", nrow(stats), "processes with data\n")
  print(as.data.frame(stats))

  # Save summary table
  write_tsv(stats, file.path(args$outdir, paste0(args$output_prefix, "_summary.tsv")))

  # Build figure (two stacked panels via patchwork; legend on bottom panel only)
  p1 <- make_panel(stats, "mean_duration", "sd_duration",
                    "Duration (hours)", "A. Runtime", show_legend = FALSE)
  p2 <- make_panel(stats, "mean_memory", "sd_memory",
                    "Peak Memory (GB)", "B. Peak Memory", show_legend = TRUE)

  fig <- p1 / p2

  # ~7.1 x 5.9 inches (matching FULL_WIDTH x STANDARD_HEIGHT from plot_utils.py)
  outpath <- file.path(args$outdir, paste0(args$output_prefix, ".png"))
  ggsave(outpath, fig, width = 7.1, height = 5.9, dpi = 300, bg = "white")
  cat("Saved", outpath, "\n")

  outpath_pdf <- file.path(args$outdir, paste0(args$output_prefix, ".pdf"))
  ggsave(outpath_pdf, fig, width = 7.1, height = 5.9, bg = "white")
  cat("Saved", outpath_pdf, "\n")

  cat("Done!\n")
}

main()
